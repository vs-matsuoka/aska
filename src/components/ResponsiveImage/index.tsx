import { css } from '@emotion/react';
import classNames from 'classnames';
import Image, { ImageProps } from 'next/image';
import { Key } from 'react';

function ResponsiveImage(
  props: Omit<ImageProps, 'fill' | 'width' | 'height'> & {
    width: number;
    height: number;
    alwaysResponsive?: boolean;
    innerKey?: Key | null;
    innerClassName?: string;
  }
) {
  const { alwaysResponsive, width, height, src, alt, className, innerKey, innerClassName, ...imageProps } = props;
  const standard = 1920;

  if (width === standard) {
    return <Image src={src} alt={alt} width={width} height={height} className={classNames(innerClassName)} {...imageProps} {...(innerKey ? { key: innerKey } : {})} />;
  }

  const responsiveImage = css({
    width: `${(width / standard) * 100}vw`,
    height: `${(height / standard) * 100}vw`,
    objectFit: 'cover',
    '@media (min-width: 1920px)': {
      width: `${width}px`,
      height: `${height}px`
    }
  });

  return (
    <div className={classNames('relative', className)} css={responsiveImage}>
      <Image src={src} alt={alt} className={classNames('max-w-none', innerClassName)} fill sizes={`${width}px`} {...imageProps} {...(innerKey ? { key: innerKey } : {})}></Image>
    </div>
  );
}

export default ResponsiveImage;
