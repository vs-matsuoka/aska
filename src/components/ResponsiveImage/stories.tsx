import { Story, Meta } from '@storybook/react';
import ResponsiveImage from '.';

export default {
  title: 'ResponsiveImage',
  component: ResponsiveImage
} as Meta;

export const Default: Story = () => <ResponsiveImage src="/Entry/21_Entry_text_02_shadow.png" width={480} height={105} alwaysResponsive alt="a" />;
