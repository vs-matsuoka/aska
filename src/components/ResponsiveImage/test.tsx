import { matchers } from '@emotion/jest';
import { render, screen } from '@testing-library/react';

import ResponsiveImage from '.';

expect.extend(matchers);

describe('<ResponsiveImage />', () => {
  it('should render the heading', () => {
    const { container } = render(<ResponsiveImage src="/Entry/21_Entry_text_02_shadow.png" width={480} height={105} alwaysResponsive alt="a" />);

    // expect(container.querySelector('div')?.style.width).toBe(`${(480 / 1920) * 100}vw`);
    // expect(container.querySelector('div')?.style.height).toBe(`${(105 / 1920) * 100}vw`);

    expect(container.firstChild).toMatchSnapshot();
  });
});
