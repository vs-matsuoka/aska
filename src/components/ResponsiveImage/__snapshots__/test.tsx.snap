// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResponsiveImage /> should render the heading 1`] = `
.emotion-0 {
  width: 25vw;
  height: 5.46875vw;
  object-fit: cover;
}

@media (min-width: 1920px) {
  .emotion-0 {
    width: 480px;
    height: 105px;
  }
}

<div
  class="relative emotion-0"
>
  <img
    alt="a"
    class="max-w-none"
    data-nimg="fill"
    decoding="async"
    loading="lazy"
    sizes="480px"
    src="/_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=3840&q=75"
    srcset="/_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=16&q=75 16w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=32&q=75 32w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=48&q=75 48w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=64&q=75 64w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=96&q=75 96w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=128&q=75 128w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=256&q=75 256w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=384&q=75 384w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=640&q=75 640w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=750&q=75 750w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=828&q=75 828w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=1080&q=75 1080w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=1200&q=75 1200w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=1920&q=75 1920w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=2048&q=75 2048w, /_next/image?url=%2FEntry%2F21_Entry_text_02_shadow.png&w=3840&q=75 3840w"
    style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
  />
</div>
`;
